dotnet/sdk-phase-1.md

```diff
@@ -0,0 +1,1052 @@
+# .NET SDK - Phase 1
+
+The .NET SDK will be implemented in three loosely defined phases:
+
+* Phase 1 - Initial project, activity worker, and client
+* Phase 2 - Workflow worker
+* Phase 3 - Deterministic workflow sandbox and/or static analyzer
+
+This proposal is for phase 1.
+
+Notes for this proposal:
+
+* Intentionally loosely defined, everything subject to change
+* Some decisions are justified at the end
+* Lots of the code is not valid C# but has things like `/*...*/` for brevity or just empty implementations
+
+## Overview
+
+The .NET SDK will be backed by [sdk-core](https://github.com/temporalio/sdk-core), utilizing "P/Invoke" and a
+custom-built bridge.
+
+The high-level goals for the .Net SDK are:
+
+* Be as simple as possible from a caller POV
+* Be as lightweight as reasonable with regards to dependencies
+* Be reasonably flexible with reasonable defaults
+* Fit well into the .Net ecosystem
+* Minimum versions supported: .NET Core 3.1, .NET Standard 2.0, and .NET Framework 4.6.2
```

Comment:
Just for my own understanding how did we choose these versions?

Reply:
It looks like .NET core 3.1 will be unsupported by MSFT in just a couple of days. [Spurce](https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-core)

Reply:
:+1: .NET core 3.1 support doesn't significantly harm us though.

Reply:
Not specifically pushing for more recent versions, but maybe we should set a soft goal and a hard goal... I mean, I don't see any reason to go out of our way to support .Net core 3.1 or Framework 4.6.2, but as long as they don't make things more complicated than they need to, then keeping them would be nice.

Reply:
Question: This proposal is about phase 1 (client + activity worker), which should be not too hard to make compatible across .Net editions. However, I expect that workflow workers, and particularly sandboxing, might require going more in depth on language features, making it much more difficult to implement on some of these. Do we have some insights on this at this point?

Reply:
> Not specifically pushing for more recent versions, but maybe we should set a soft goal and a hard goal... I mean, I don't see any reason to go out of our way to support .Net core 3.1 or Framework 4.6.2, but as long as they don't make things more complicated than they need to, then keeping them would be nice.

Yeah, not wanting to go out of our way. This is what https://www.nuget.org/packages/Microsoft.Azure.WebJobs.Extensions.DurableTask, a similar library, supports. But you can't have "soft" and "hard", basically you need to have CI for the lowest and most recent at least.

> Do we have some insights on this at this point?

I do not. I have not even put any worthwhile research into what .NET sandboxing approaches may exist for us.

Reply:
>But you can't have "soft" and "hard", basically you need to have CI for the lowest and most recent at least.

Sure... We can start by setting the CI to number X, but know that if you face any difficulty that can be resolved by updating to X+1, then you just do so, without remorse and without loosing a week researching/demonstrating/arguing about how it would be possible to get around that in version X...